# Homework


## Submit the results

* Submit your results here: https://forms.gle/jneGM91mzDZ23i8HA
* You can submit your solution multiple times. In this case, only the last submission will be used 
* If your answer doesn't match options exactly, select the closest one


## Deadline

The deadline for submitting is 18 September 2022 (Monday), 23:00 CEST (Berlin time).

After that, the form will be closed.


### Question 1

What's the version of Pandas that you installed?

You can get the version information using the `__version__` field:

```python
pd.__version__
```

In [4]:
import pandas as pd
pd.__version__

'1.4.4'

### Getting the data 

For this homework, we'll use the California Housing Prices dataset. Download it from 
[here](https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv).

In [5]:
df = pd.read_csv('raw.githubusercontent.com_alexeygrigorev_datasets_master_housing.csv')
df

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  
0            NEAR BAY  
1            NEAR BAY  
2            NEAR BAY  
3            NEAR BAY  
4            NEAR BAY  
...               ...  
20635          INLAND  
20636          INLAND  
20637          INLAND  
20638          INLAND  
20639          INLAND  

[20640 rows x 10 columns]

### Question 2

How many columns are in the dataset?

- 10
- 6560
- 10989
- 20640

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


There are 10 columns in the dataset.

### Question 3

Which columns in the dataset have missing values?

- `total_rooms`
- `total_bedrooms`
- both of the above
- no empty columns in the dataset

`total_bedrooms` column has missing values

### Question 4

How many unique values does the `ocean_proximity` column have?

- 3
- 5
- 7
- 9

In [10]:
df.ocean_proximity.unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

`ocean_proximity` column has 5 unique values

### Question 5

What's the average value of the `median_house_value` for the houses located near the bay?

- 49433
- 124805
- 259212
- 380440

In [14]:
df_near_bay = df[df.ocean_proximity == 'NEAR BAY']
df_near_bay

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
19067    -122.45     38.28                20.0       3306.0           503.0   
19068    -122.47     38.29                14.0       3732.0           846.0   
19069    -122.45     38.27                25.0       5024.0           881.0   
19077    -122.49     38.27                 8.0       5092.0           988.0   
19104    -122.51     38.17                 8.0       5875.0          1115.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
19067      1374.0       460.0         5.7984            297600.0   
19068      1277.0       775.0         2.5658            208000.0   
19069      1994.0       838.0         4.2237            262300.0   
19077      1657.0       936.0         3.5625            213200.0   
19104      2808.0      1029.0         3.6392            246300.0   

      ocean_proximity  
0            NEAR BAY  
1            NEAR BAY  
2            NEAR BAY  
3            NEAR BAY  
4            NEAR BAY  
...               ...  
19067        NEAR BAY  
19068        NEAR BAY  
19069        NEAR BAY  
19077        NEAR BAY  
19104        NEAR BAY  

[2290 rows x 10 columns]

In [15]:
df_near_bay.describe()

longitude     latitude  housing_median_age   total_rooms  \
count  2290.000000  2290.000000         2290.000000   2290.000000   
mean   -122.260694    37.801057           37.730131   2493.589520   
std       0.147004     0.185434           13.070385   1830.817022   
min    -122.590000    37.350000            2.000000      8.000000   
25%    -122.410000    37.730000           29.000000   1431.250000   
50%    -122.250000    37.790000           39.000000   2083.000000   
75%    -122.140000    37.907500           52.000000   3029.750000   
max    -122.010000    38.340000           52.000000  18634.000000   

       total_bedrooms   population   households  median_income  \
count     2270.000000  2290.000000  2290.000000    2290.000000   
mean       514.182819  1230.317467   488.616157       4.172885   
std        367.887605   885.899035   350.598369       2.017427   
min          1.000000     8.000000     1.000000       0.499900   
25%        289.000000   718.250000   275.000000       2.834750   
50%        423.000000  1033.500000   406.000000       3.818650   
75%        628.750000  1495.000000   599.250000       5.054425   
max       3226.000000  8276.000000  3589.000000      15.000100   

       median_house_value  
count         2290.000000  
mean        259212.311790  
std         122818.537064  
min          22500.000000  
25%         162500.000000  
50%         233800.000000  
75%         345700.000000  
max         500001.000000

The average value of the `median_house_value` for the houses located near the bay is 259212

### Question 6

1. Calculate the average of `total_bedrooms` column in the dataset.
2. Use the `fillna` method to fill the missing values in `total_bedrooms` with the mean value from the previous step.
3. Now, calculate the average of `total_bedrooms` again.
4. Has it changed?

Has it changed?

> Hint: take into account only 3 digits after the decimal point.

- Yes
- No

In [16]:
df.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.569704     35.631861           28.639486   2635.763081   
std        2.003532      2.135952           12.585558   2181.615252   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20640.000000  20640.000000   20640.000000   
mean       537.870553   1425.476744    499.539680       3.870671   
std        421.385070   1132.462122    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    787.000000    280.000000       2.563400   
50%        435.000000   1166.000000    409.000000       3.534800   
75%        647.000000   1725.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

In [17]:
# Calculate the mean of total_bedrooms
mean_total_bedrooms = df['total_bedrooms'].mean()
mean_total_bedrooms

537.8705525375618

In [18]:
# Fill missing values in total_bedrooms with the calculated mean
df['total_bedrooms'].fillna(mean_total_bedrooms, inplace=True)

In [19]:
# Calculate the mean of total_bedrooms after filling average
mean_total_bedrooms = df['total_bedrooms'].mean()
mean_total_bedrooms

537.8705525375639

The average of `total_bedrooms` hasn't changed after taking into account only 3 digits after the decimal point.

### Question 7

1. Select all the options located on islands.
2. Select only columns `housing_median_age`, `total_rooms`, `total_bedrooms`.
3. Get the underlying NumPy array. Let's call it `X`.
4. Compute matrix-matrix multiplication between the transpose of `X` and `X`. To get the transpose, use `X.T`. Let's call the result `XTX`.
5. Compute the inverse of `XTX`.
6. Create an array `y` with values `[950, 1300, 800, 1000, 1300]`.
7. Multiply the inverse of `XTX` with the transpose of `X`, and then multiply the result by `y`. Call the result `w`.
8. What's the value of the last element of `w`?

> **Note**: You just implemented linear regression. We'll talk about it in the next lesson.

- -1.4812
- 0.001
- 5.6992
- 23.1233

In [21]:
df_island = df[housing_median_age == 'ISLAND']
df_island

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
8314    -118.32     33.35                27.0       1675.0           521.0   
8315    -118.33     33.34                52.0       2359.0           591.0   
8316    -118.32     33.33                52.0       2127.0           512.0   
8317    -118.32     33.34                52.0        996.0           264.0   
8318    -118.48     33.43                29.0        716.0           214.0   

      population  households  median_income  median_house_value  \
8314       744.0       331.0         2.1579            450000.0   
8315      1100.0       431.0         2.8333            414700.0   
8316       733.0       288.0         3.3906            300000.0   
8317       341.0       160.0         2.7361            450000.0   
8318       422.0       173.0         2.6042            287500.0   

     ocean_proximity  
8314          ISLAND  
8315          ISLAND  
8316          ISLAND  
8317          ISLAND  
8318          ISLAND

In [23]:
columns_of_interest = ['housing_median_age', 'total_rooms', 'total_bedrooms']
columns_of_interest

['housing_median_age', 'total_rooms', 'total_bedrooms']

In [25]:
df_island = df_island[columns_of_interest]
df_island

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

In [26]:
import numpy as np


X = df_island.to_numpy()
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [27]:
XTX = np.dot(X.T, X)
XTX

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [28]:
XTX_inverse = np.linalg.inv(XTX)
XTX_inverse

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [29]:
y = np.array([950, 1300, 800, 1000, 1300])
y

array([ 950, 1300,  800, 1000, 1300])

In [30]:
w = np.dot(np.dot(XTX_inverse, X.T), y)
w

array([23.12330961, -1.48124183,  5.69922946])

In [31]:
last_element_of_w = w[-1]
last_element_of_w

5.699229455065575

The value of the last element of `w` is 5.6992